### DISCUSSION TITLE
How to deal with evolving schemas?

### DISCUSSION THREAD

**Discussion by o-smirnov** [caracal-pipeline/cult-cargo#20] [Ideas]
As raised by @JSKenyon here https://github.com/caracal-pipeline/cult-cargo/pull/18. But of course this problem is not unique to QuartiCal or pfb-clean. Wsclean also likes to introduce new parameters with new versions. How to deal with this cleanly?

**Comment by o-smirnov** [comment]
Just nothing that there's two problems to address:

* How to propagate schema files into ``cult-cargo`` **without** making QuartiCal or pfb-clean an explicit dependency.

* How to provide multiple schema versions within ``cult-cargo``. The brute-force way is to propagate version names into the cab name (so e.g. the cab name becomes ``quartical-0.2.2``, while ``quartical`` itself refers to perhaps the newest version). A more elaborate way would be to add a ``versions`` field to the parameter definition where one could specify, somehow, which version a particular parameter is valid for.

**Reply by o-smirnov** [reply]
A third, less invasive way perhaps: provide different yaml files for different versions. E.g. ``quartical-0.2.2.yaml`` could contain
```yml
cabs:
  quartical-0.2.2:
    image:
      version: 0.2.2-{vars.cult-cargo.images.version}
     ...
  quartical:
    _use: quartical-0.2.2
```
and ``quartical.yaml`` itself could contain just an _include of the "latest" version at time of cult-cargo release.

This way the end-user looking for a particular version would only need to change the name of the include file, while the body of the recipe could continue referring to ``quartical`` generically.

**Reply by JSKenyon** [reply]
The matter is even more complicated when you consider other applications. QC and PFB fundamentally speak the same language as stimela. `wsclean` doesn't have a cab definition in it i.e. I think one would always have to do this by hand. One thing to consider is doing something like conda. They have a [bot](https://github.com/regro/cf-scripts) which seems to automatically notice releases and makes an appropriate PR into conda feedstock (@sjperkins thoughts?). That might be something we can replicate, although I am not sure I would know where to begin. We could also consider something triggering actions on cult-cargo using [this approach](https://stackoverflow.com/questions/58465057/trigger-a-github-action-when-another-repository-creates-a-new-release).

**Reply by sjperkins** [reply]
Apologies if this has already been suggested elsewhere, but probably what we're thinking of is a registry of cab definitions.

Depending on the version of the running application, i.e., looked up with something like `docker run --rm goquartical --version`, the appropriate cab yaml could be discovered in the registry and applied. I'm mostly thinking of this in the context of vcpkg which maintains a large registry of cmake glue for installing c++ depedencies: https://github.com/microsoft/vcpkg/tree/master/ports

One easy way to do this might be to just have a git registry with the appropriate yamls, each for a different application version.

This might not be too different from prior iterations of stimela.

Of course, new software releases would need to submit updated schemas to the registry. It might be possible to automate these as @JSKenyon suggested.

**Reply by JSKenyon** [reply]
To be clear though, I am not quite sure if it would be possible to avoid updating cab definitions by hand for applications not using stimela as a parser. At the very least, as mentioned in my other comment, having PRs opened automatically on release/tag would at least make keeping track a little simpler.

**Reply by o-smirnov** [reply]
Yeah no avoiding doing this by hand for third-party apps, this is true. But at least we can come up with the most labour-saving way to do this for the ones we control...

**Reply by sjperkins** [reply]
I did some quick googling for `yaml registry schemas`, but notihing OSS and obvious popped up. I'm also certain someone else has had to think of this problem before (k8s was one of the search results).

Perhaps with more digging there wouldn't be a need to reinvent the wheel.

**Reply by o-smirnov** [reply]
But do we even need to invent this wheel? In a sense, the cult-cargo repo **is** the schema registry already. A new release of an app comes along, we push an update and issue an "rc" release for the brave. Seems simple enough, we just need a smoother way to copy the yamls over.

**Reply by sjperkins** [reply]
Maybe not, I know very little about cult-cargo, but it sounds like it could be repurposed.

FWIW vcpkg keeps a versions directory for different `ports` (cabs?). For example fftw3:

https://github.com/microsoft/vcpkg/blob/master/versions/f-/fftw3.json

These in turn refer to different git commits referencing versions of the `port` (cab?) in this directory

https://github.com/microsoft/vcpkg/tree/master/ports/fftwpp

**Comment by JSKenyon** [comment]
Copied from #18. I think that the correct thing to do is to have a GitHub action that handles this automatically. This could be on either the application repo or on cult-cargo. The latter would be preferable but the former is probably easier. Basically, on release, the action should pick out the required files (a reason to have this on the application repos) and automatically make a PR into cult-cargo pushing those files to some folder based on the current release version. This means that the application developer only has to worry about it once, and it simply becomes merging a PR on the cult-cargo end.

**Reply by o-smirnov** [reply]
Yeah I like that idea. Make it an action on the application repo side, using a script that we can provide as part of ``cult-cargo``. We would still need to take some manual action on the ``cult-cargo`` side (updating the build manifest, for instance), so opening a PR automatically is the perfect way to prompt this.

**Reply by landmanbester** [reply]
There is still something I don't like about this but I am not sure of a way to get around it viz. it requires a new cult-cargo release to reflect changes on the application side. It may be preferable to handle this during installation of cult-cargo since that way the change will be reflected upon reinstallation of cult-cargo

**Reply by JSKenyon** [reply]
I see your point. If we need to get changes into the hands of the user as quickly as possible, it will be irritating having to wait for cult-cargo releases. The alternative is to make the version registry something that exists outside of the release and that gets checked at runtime. That way cult-cargo releases could be decoupled from simple changes to config schemas/application releases. I think we can go a step further and not even worry about reinstalling cult-cargo, provided that information can easily be pulled in at run-time.

**Reply by landmanbester** [reply]
Even better. Not sure how these registries work but if they could refer to the config files in the application repo that would be first prize since that avoids the need to have multiple copies of the config files

**Reply by o-smirnov** [reply]
Yes but that's sort of by design. Any given release of ``cult-cargo`` captures a "snapshot" of the app versions available at release time -- it'as pretty hard to ensure reproducibility without that. In fact, it would be downright sneaky and evil if a simple reinstall of the **same** cult-cargo release suddenly changed the outcome just because a new release of pfb-clean came along in the interim.

Other arrangements can be made for those needing to live on the bleeding edge. A user can always pull schemas from his local branch of pfb-clean, for example.

As for "irritating having to wait" -- not necessarily. It's very quick and non-intrusive to push a cult-cargo "release candidate" version. These things are ignored by pip unless the version is explicitly requested with ==x.y.zrcN. So if it's only a matter of pushing a quick QuartiCal update into the hands of some testers, the rc mechanism is an easy avenue for that.

**Reply by tmolteno** [reply]
As a noob, I am not sure that schema changes would usually need to appear quickly in cult-cargo other than for developers of new cabs. In a sense cult-cargo seems like a linux distribution. How about following the tried and tested: stable, testing and unstable and adopt a release cycle?

**Reply by JSKenyon** [reply]
> Yes but that's sort of by design. Any given release of `cult-cargo` captures a "snapshot" of the app versions available at release time -- it'as pretty hard to ensure reproducibility without that. In fact, it would be downright sneaky and evil if a simple reinstall of the **same** cult-cargo release suddenly changed the outcome just because a new release of pfb-clean came along in the interim.

Oddly, I think this argues against the `latest` form of image specification. As you point out, it means that your pipeline will only be reproducible with a specific version of `cult-cargo`. Shouldn't the aim be to reproduce any pipeline, regardless of the cult-cargo version? In other words, given the latest version of `cult-cargo` it should always have sufficient information to give you the exact versions which were used. 

> As for "irritating having to wait" -- not necessarily. It's very quick and non-intrusive to push a cult-cargo "release candidate" version. These things are ignored by pip unless the version is explicitly requested with ==x.y.zrcN. So if it's only a matter of pushing a quick QuartiCal update into the hands of some testers, the rc mechanism is an easy avenue for that.

Technically we could be like `casadata` and do daily calver releases.

**Reply by o-smirnov** [reply]
> Oddly, I think this argues against the latest form of image specification. As you point out, it means that your pipeline will only be reproducible with a specific version of cult-cargo. 

I think that's the best that can be achieved, realistically. The best we can hope for, given any pipeline, is a statement like "this can be reproduced with wsclean 3.0.1, QuartiCal 0.2.2 and CASA 6.5.7 and and and ...". A "specific version of cult-cargo" is just a shorthand way of snapshotting the zoo of the underlying software versions.

Which why I don't mind the "latest" tag -- which currently has the strict meaning of "latest as per this version of cult-cargo", and therefore also refers to a static, non-evolving version. 

At the end of the day, reproducibility requires that somebody keep track of all the version numbers used. Having everything captured in a cult -cargo release takes the onus off the user, at least.

> Shouldn't the aim be to reproduce any pipeline, regardless of the cult-cargo version?

I don't think that's possible in general -- see arguments above.

**Reply by tmolteno** [reply]
As a noob, I agree. It might be a question of philosophy. Does the  Curated in "Curated Stimela2 cargo for popular radio astronomy software" mean that the cargo are more likely to play nicely with each other?

**Reply by JSKenyon** [reply]
I think I disagree at some level. This means that a recipe will be tightly coupled to `cult-cargo` i.e. if I come along and read the recipe, I may have literally no idea what versions were used if they were all simply `latest` at that point. To be clear, I am not saying that that means it isn't reproducible, I am merely arguing that it completely obfuscates versions which in turn makes high level debugging difficult. Imagine that I knew there was a hideous bug in `quartical==0.2.2` (hopefully there isn't). By the above logic, if someone sent me their recipe I would have no clue if that was the version they ultimately used without first getting the appropriate version of `cult-cargo` and figuring out what `latest` was at the time. I am playing Devil's advocate here; I appreciate the simplicity of `latest` but I also feel that explicit is better than implicit when it comes to versioning.

**Reply by landmanbester** [reply]
I may be mistaken but I believe one of the biggest hurdles for users of old stimela was waiting for new releases to propagate changes in the underlying packages. I thought the whole idea behind the way we have set up applications in the new ecosystem (at this point really just QC and PFB) is for the config rendering etc. to happen independently of stimela/cult-cargo. The fact that we need to explicitly duplicate config files in cult-cargo does not align well with this idea so I think it may be worth giving this some more thought. It would be ideal if the configs live in one place only and we can somehow link to them at runtime. I will add that I don't think reproducibility can or should be guaranteed if using the ```latest``` form of cab specification, only if the versions are fixed

**Reply by tmolteno** [reply]
I agree with latest being dangerous. Very little harm comes from explicit versioning, and it avoids some nasty surprises.

**Reply by o-smirnov** [reply]
Well, there's an essential tension between "ease of use", "seamless access to latest version", and "reproducibility", and I'm not sure we can satisfy all three. For example:

> I think I disagree at some level. This means that a recipe will be tightly coupled to cult-cargo i.e. if I come along and read the recipe, I may have literally no idea what versions were used if they were all simply latest at that point.

But then the onus is on the user to explicitly specify versions at every step. Which compromises the ease of use substantially, and clutters the recipes. The "latest" tag is meant to relieve this drudgery -- just tell us which cult-cargo you're using, and we can take it from there. (I'm thinking of cult-cargo as the KERN of stimela...)

> It would be ideal if the configs live in one place only and we can somehow link to them at runtime. 

It is possible to extend configuratt's _include statement to support a URL, so e.g. to pull the schema directly off the pfb-clean repo. But then this undermines reproducibility -- and requires a runtime internet connection, which clusters don't always grant you. Still might be a handy option to have for the bleeding-edge people. But a "reproducible" recipe must, by necessity, come with a fixed snapshot of the schemas (and docker images) used by that recipe. A "specific release of cult-cargo" provides exactly that snapshot -- but maybe there's an even easier way to do it which we just haven't figured out yet.

**Reply by JSKenyon** [reply]
As an aside, I would then advocate for a `stimela freeze` argument that either dumps the versions used (I think you actually already have this or something like it) or goes and explicitly updates them in the recipe. In fact, that appeals. The workflow would be:

1. Write a recipe using a given version of `cult-cargo` and `latest` tags.
2. It works! Hooray!
3. Run `stimela freeze recipe.yaml` which dumps out the recipe with all the `latest` tags replaced with explicit versions.

How difficult/useful do you think that would be Oleg? I feel like it would be a more robust thing to pass around. It also opens the door to decoupling your recipe from a specific `cult-cargo` version.

**Reply by tmolteno** [reply]
It would be nice  when reading recipes to know which version of cult-cargo was used then.  How about a syntax that said:

```
# include some of my cab definitions
_include:
  - (cultcargo:10)wsclean.yml
  - (cultcargo:10)tricolour.yml
```

And use release numbering like debian or ubuntu

**Reply by o-smirnov** [reply]
Also, bear in mind that we could impose "rules", or at least guidelines, for certified-reproducible workflows. I.e. "thou shalt explicitly specify all package versions used" is necessary -- I'm merely arguing that "thou shalt specify the cult-cargo used, and then we can let you get away with 'latest'" is a less onerous guideline which achieves the same result.

In the meantime, the development workflows are free to play with the more fancy features.

**Reply by o-smirnov** [reply]
> As an aside, I would then advocate for a stimela freeze argument that either dumps the versions used (I think you actually already have this or something like it)

Yeah, the ``stimela.deps`` file being dumped was a start towards something like this. I haven't seen my way through it completely -- bear in mind it's not just cab versions, it's all the other YaML being included that needs freezing. Of course, a brute-force solution is to have the freeze command pump out one massive YaML document with all the includes applied. That would be trivial to implement -- it just won't make for pleasant reading. But it's the ultimate in robustness, as that single YaML doc is completely self-contained.

I think @JSKenyon is suggesting a halfway house where the recipe is augmented with cab versions. Note that this doesn't even require changing the recipe. Just scan the thing and dump out a separate YaML file containing the versions which can be merged in via a _post_include statement (or just by specifying that extra file on the command line).

> It would be nice when reading recipes to know which version of cult-cargo was used then. How about a syntax that said:

```
_include:
  - (cultcargo:10)wsclean.yml
```

Interesting idea... well you can't really *select* a version, since only one version is installed in the virtualenv, but you can treat it as a pip-style version requirements specifier I guess, something like ``(cultcargo==0.1.2)`` or ``(cultcargo>0.1.2)``. At least then stimela can bomb out if the version requirements are not met.

**Reply by tmolteno** [reply]
Yep that would be fine for me (https://peps.python.org/pep-0440/#version-specifiers).

**Reply by landmanbester** [reply]
> It is possible to extend configuratt's _include statement to support a URL, so e.g. to pull the schema directly off the pfb-clean repo. But then this undermines reproducibility -- and requires a runtime internet connection, which clusters don't always grant you. Still might be a handy option to have for the bleeding-edge people. But a "reproducible" recipe must, by necessity, come with a fixed snapshot of the schemas (and docker images) used by that recipe. A "specific release of cult-cargo" provides exactly that snapshot -- but maybe there's an even easier way to do it which we just haven't figured out yet.

I'm wondering if it would be possible to use such a mechanism to make the dev version of cult-cargo a bit more dev friendly. For example, could we not set things up so that the dev version of cult-cargo fetches config files from the application repo when using the ```latest``` form of cab specification? ```latest``` would still refer to a specific version of a cab for any cult-cargo release (something that can probably be frozen in using github actions or the freeze mechanism @JSKenyon is proposing) so recipes relying on official releases would still be reproducible but we don't guarantee reproducibility when using the the dev version of cult-cargo. This way there is only one source of truth for the config and this automatically gets propagated to cult-cargo upon release (effectively getting added to the registry) and users are also able to quickly test fixes simply by installing a dev version of cult-cargo without the need to install other packages. Of course, they would need to set their image-pull-policies to always to get up to date Docker images but this is still something that happens at the level of the config without the need for making new releases. Just thinking out loud here, not sure if this is actually feasible/practical. Thoughts?

**Reply by o-smirnov** [reply]
Maybe the dev version of ``cult-cargo`` should be a separate package called ``fresh-cargo`` -- no reason having one package that tries to be all things to all people -- and then it can have the fancy runtime config schema fetch shizzle?

**Reply by landmanbester** [reply]
```hip-cargo``` 😎 

We could probably also have ```main``` versions for this, that way ```latest``` keeps its current meaning

**Reply by tmolteno** [reply]
I like the idea of explicit versioning

- ```cult-cargo:testing```  (refer to the upcoming next stable, this is the version that will become the next release)
- ```cult-cargo:2024.02``` (refer to an explicit release)
- ```cult-cargo``` without versioning should throw an error

Using a different package called ```fresh-cargo```: love it, and makes things easier (as an alternative to the first option above). Another name might be ```unstable-cargo``` to hint to the user the gravity of using this in production scripts.
